### PR TITLE
[Cosmos] Bulk api implementation

### PR DESCRIPTION
**Pull request by simorenoh**
These changes add Bulk requests to our SDK.
image

I have updated the README to show what these look like, but the operations that get sent to the method look like this:
image

We have added the first models to our PythonSDK responses with this PR as well. Bulk responses will be returned in a list made up of OperationResult and OperationError objects:
image
image
 
From the meeting we had with the Core SDK, we have also addressed:
- Removed the response headers from the response and now instead we are saving the aggregated headers into the client's last response headers, making it align with the way it is used in our other methods
- Added tests for single item size limit with bulk - as expected, the service returns an error for the given item, but tests have been included regardless. Have also ran some tests to verify unicode characters are non-problematic with the size of the batches.
- Included more information on using request timeouts with bulk and added this documentation to the readme section on bulk. in this case since each batch is sent separately, the request timeout is applied on a per-batch basis as opposed to an entire bulk request basis, so I've made sure to make that clear to users in this section so they know how this works.
As part of this however, I had a question - since this only applies to each batch within the bulk request, would it make sense for us to continue with the pending batches that haven't ran? or should we cancel them as well?
- Return type for bulk operations is now a Union[List[]] type like mentioned above

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py [hunk 1]
```diff
@@ -1765,6 +1769,83 @@ def PatchItem(self, document_link, operations, options=None, **kwargs):
...
+                operation = self._AddPartitionKey(collection_link, operation.get("resourceBody"), operation)
+                partition_key_value = operation.get("partitionKey", None)
+                if partition_key_value is None:
+                    raise AttributeError("Bulk operation must have a partition key specified.")
+            operation["partitionKey"] = json.dumps([partition_key_value])
+            pk_definition = self._get_partition_key_definition(collection_link)
+            container_pk = partition_key.PartitionKey(path=pk_definition.get('path'),
+                                                      kind=pk_definition.get('kind'),
+                                                      version=pk_definition.get('version'))
+            hashed_key = container_pk.get_effective_partition_key_string([partition_key_value])
```
Comment (NaluTripician) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py:1811]: Just a note here, in .NET we ran into some trouble using the effective partition key string over an effective partition key range with queries for subpartitioning with prefix partition key. Since this is bulk and not query it might be a non-issue here.

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py [hunk 2]
```diff
@@ -623,6 +623,39 @@ def patch_item(
...
+        self,
+        operations: List[Dict[str, Any]],
+        **kwargs: Any
+    ) -> Dict[str, Any]:
+        """ **Provisional method** Bulk execute the list of operations on the given items.
+        Can be used for create, upsert, read, replace and delete operations.
+
+        :param operations: The list of bulk operations to execute.
+        :type operations: List[Dict[str, Any]]
+        :keyword str pre_trigger_include: trigger id to be used as pre operation trigger.
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/azure/cosmos/container.py:637]: Do we absolutely need triggers for bulk?

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
I would like us to discuss how we think we should expose bulk-support in public API in the Python SDK.

It looks like the current proposal kind of mimics the API in the Java SDK. The Java SDK has intentionally very low-level APIs (because we also needed to make sure we can use them in Spark- and Kafka Connector). What is still missing in the Java SDK is a higher level abstraction API - we have a sample here - https://github.com/Azure-Samples/azure-cosmos-java-sql-api-samples/blob/main/src/main/java/com/azure/cosmos/examples/bulk/async/BulkWriter.java

But even that has a few gaps.

Even on the Java side we have seen quite a number of CRIs asking for how to use the Bulk API - even when usually our Java developers using direct mode in Cosmos are kind of used to a lower level abstraction interaction with Cosmos DB.

IMO we should learn from this experience - and when designing Bulk for Gateway-only SDKS (python, JS, Go), we should expose public API that has a higher abstraction level.

What do I mean by that?
The API we expose should handle the following challenges for every bulk workload without the caller/user having to deal with all the details (we might still need some policy level configs to influence certain things)
- Throughput guard-rails / Batch size - This is something that has really helped us in the Java SDK - being to dynamically tune the micro batch size sent to the service within one request (especially in combination with client throughput control). We won't have throughput control in Python anytime soon - but we still need to have some way for customers to ensure a single machine doing bulk ingestion won't consume all RU/s -  otherwise a series of CRI will just come out of that. Could be wait intervals and dynamically tuning batch size or similar, can be more naive than in Java - but it needs to set some guard-rails
- Retries - current proposal is doing retries for 429 and 410 - goodness, but not sufficient - what about 449, request timeouts etc. Also, I would really look at seeing whether we can expose public API that forces bulk operations to be idempotent - look at the Spark connector and the ItemWriteStrategy - it automatically ensures that all write operations are idempotent - so, we can fully own retries - even on transit timeouts - and it can also allow policy driven handling of 409/412

To me the two points above are really critical. In the vast majority of real-world use cases, the (intentional) bottle-neck for bulk/batch scenarios is the service provisioned throughput. It is relatively easy to saturate the service-side throughput. So, building a bulk interface means optimizing for how to best ensure that provisoned throughput on the backend (or whatever "portion" the customer wants to use for bulk) is saturated ideally all the time - which defeats the retry solution in the current proposal - if you have a batch of 100 items and you wait until all of them are finished before picking up the next batch you are gated behind the long-pole physical partition - before you pick up the next batch. This will result in not being able to saturate the throughput on the other partitions.

Parts of the work in this PR will be helpful as an internal API for sure - but when we look at the API we want to expose to devs, I think in the Gateway SDKs we should choose a higher abstraction level - and we should add similar high-level APIs eventually in Java as well. Otherwise the support cost will simply be too high.

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py (hunk 2 above)
Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/azure/cosmos/container.py:637]: every request in the Python SDK exposes triggers to users, it's part of the Python SDK designs by Core team

**Comment by simorenoh** [comment]
Closing this for now since we need further design and conversation on throughput guardrails and current behavior.
